Introduction of CUDA Graphs to LLama.cpp

agray3 opened the pull request:
See Issue #6763

agray3 committed:
DRAFT: Introduction of CUDA Graphs to LLama.cpp

agray3 commented:
Labelled as DRAFT since this will need some more testing across different models, CUDA versions, etc before it is merged. See https://github.com/ggerganov/llama.cpp/issues/6763.

phymbert commented:
@ggerganov please restart the CI github manager for benchmark

JohannesGaessler commented:
`./llama-bench` and `./perplexity` are broken with this PR which is why I'm using `./main` for testing. Results for LLaMA 2 7b q4_0:

| GPU         | Test      | t/s master | t/s PR | Speedup |
|-------------|-----------|------------|--------|---------|
| 1x RTX 3090 | ./main pp |       3445 |   3185 |    0.92 |
| 1x RTX 3090 | ./main tg |     116.31 | 128.27 |    1.10 |
| 1x RTX 4090 | ./main pp |       6668 |   6963 |    0.96 |
| 1x RTX 4090 | ./main tg |     138.31 | 146.81 |    1.06 |

The speedup does not seem to be consistent: for a batch size of 1 it's faster but for batch sizes >> 1 it's slower. Still, the potential speedup is higher than I thought so it seems I was wrong when I previously said using CUDA graphs would not be worthwhile.

<details>
<summary>Error for perplexity</summary>

```
johannesg@johannes-romed82t-00 ~/Projects/llama.cpp                                                                         [15:04:17]
> $ ./perplexity --file wikitext-2-raw/wiki.test.raw --model models/opt/${model_name}-${quantization}.gguf -ngl 99 -c 4096
main: build = 2699 (cec409aa)
main: built with cc (GCC) 13.2.1 20230801 for x86_64-pc-linux-gnu
main: seed  = 1713531858
llama_model_loader: loaded meta data with 22 key-value pairs and 291 tensors from models/opt/llama_2-7b-q4_0.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = LLaMA v2
llama_model_loader: - kv   2:                       llama.context_length u32              = 4096
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 11008
llama_model_loader: - kv   6:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   7:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   8:              llama.attention.head_count_kv u32              = 32
llama_model_loader: - kv   9:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                          general.file_type u32              = 2
llama_model_loader: - kv  11:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  12:                      tokenizer.ggml.tokens arr[str,32000]   = ["<unk>", "<s>", "</s>", "<0x00>", "<...
llama_model_loader: - kv  13:                      tokenizer.ggml.scores arr[f32,32000]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  14:                  tokenizer.ggml.token_type arr[i32,32000]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  15:                      tokenizer.ggml.merges arr[str,61249]   = ["▁ t", "e r", "i n", "▁ a", "e n...
llama_model_loader: - kv  16:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  17:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  18:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  19:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  20:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  21:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type q4_0:  225 tensors
llama_model_loader: - type q6_K:    1 tensors
llm_load_vocab: special tokens definition check successful ( 259/32000 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 32000
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: n_ctx_train      = 4096
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 32
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 1
llm_load_print_meta: n_embd_k_gqa     = 4096
llm_load_print_meta: n_embd_v_gqa     = 4096
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 11008
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 10000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 4096
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = Q4_0
llm_load_print_meta: model params     = 6.74 B
llm_load_print_meta: model size       = 3.56 GiB (4.54 BPW) 
llm_load_print_meta: general.name     = LLaMA v2
llm_load_print_meta: BOS token        = 1 '<s>'
llm_load_print_meta: EOS token        = 2 '</s>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: LF token         = 13 '<0x0A>'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4090, compute capability 8.9, VMM: yes
llm_load_tensors: ggml ctx size =    0.30 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =    70.31 MiB
llm_load_tensors:      CUDA0 buffer size =  3577.56 MiB
..................................................................................................
llama_new_context_with_model: n_ctx      = 4096
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =  2048.00 MiB
llama_new_context_with_model: KV self size  = 2048.00 MiB, K (f16): 1024.00 MiB, V (f16): 1024.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     0.49 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   296.00 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    16.01 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 2

system_info: n_threads = 64 / 128 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | 
perplexity: tokenizing the input ..
perplexity: tokenization took 958.351 ms
perplexity: calculating perplexity over 81 chunks, n_ctx=4096, batch_size=2048, n_seq=1
CUDA error: invalid argument
  current device: 0, in function ggml_backend_cuda_graph_compute at ggml-cuda.cu:2515
  cuGraphKernelNodeGetParams_v2(nodes[i], &paramsDriver[i])
GGML_ASSERT: ggml-cuda.cu:60: !"CUDA error"
[1]    1644936 IOT instruction (core dumped)  ./perplexity --file wikitext-2-raw/wiki.test.raw --model  -ngl 99 -c 4096
```

</details>

agray3 commented:
Thanks for these tests. I haven't yet optimized/tested for batch size greater than one - it might be a good idea for me to only enable CUDA graphs for size 1 initially. I'll also look at the failures.
>it seems I was wrong when I previously said using CUDA graphs would not be worthwhile.
It's not obvious - even without CUDA graphs, llama.cpp already does a good job of pre-launching all kernels in the GGML graph, so CPU-side launch overheads are not the issue. But CUDA graphs also optimises GPU-side launch overheads, to reduce the "gaps" between kernels, and that is the benefit we are seeing here.

sorasoras commented:
```
.\llama-bench.exe -m W:\model\mistral-7b-v0.1.Q5_K_S.gguf -ngl 99 -sm none
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 2 CUDA devices:
  Device 0: Tesla P40, compute capability 6.1, VMM: yes
  Device 1: Tesla P40, compute capability 6.1, VMM: yes
| model                          |       size |     params | backend    | ngl | sm         | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------- | ---------------: |
```
It does not seems to work at P40.

and I cannot get it compile on ROCM

```
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2410:5: error: unknown type name 'cudaGraph_t'
    cudaGraph_t graph = nullptr;
    ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2411:5: error: unknown type name 'cudaGraphExec_t'; did you mean 'hipGraphExec_t'?
    cudaGraphExec_t instance = nullptr;
    ^~~~~~~~~~~~~~~
    hipGraphExec_t
C:/Program Files/AMD/ROCm/5.7/include\hip/hip_runtime_api.h:1098:30: note: 'hipGraphExec_t' declared here
typedef struct hipGraphExec* hipGraphExec_t;
                             ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2458:63: error: use of undeclared identifier 'cudaStreamCaptureModeGlobal'
        CUDA_CHECK(cudaStreamBeginCapture(cuda_ctx->stream(), cudaStreamCaptureModeGlobal));
                                                              ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2501:9: error: unknown type name 'cudaGraphNode_t'; did you mean 'hipGraphNode_t'?
        cudaGraphNode_t nodes[MAX_NODES_IN_CUDA_GRAPH];
        ^~~~~~~~~~~~~~~
        hipGraphNode_t
C:/Program Files/AMD/ROCm/5.7/include\hip/hip_runtime_api.h:1094:30: note: 'hipGraphNode_t' declared here
typedef struct hipGraphNode* hipGraphNode_t;
                             ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2502:9: error: unknown type name 'CUDA_KERNEL_NODE_PARAMS_v2'
        CUDA_KERNEL_NODE_PARAMS_v2 paramsDriver[MAX_NODES_IN_CUDA_GRAPH];
        ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2503:9: error: unknown type name 'cudaKernelNodeParams'; did you mean 'hipKernelNodeParams'?
        cudaKernelNodeParams paramsRuntime[MAX_NODES_IN_CUDA_GRAPH];
        ^~~~~~~~~~~~~~~~~~~~
        hipKernelNodeParams
C:/Program Files/AMD/ROCm/5.7/include\hip/hip_runtime_api.h:1139:3: note: 'hipKernelNodeParams' declared here
} hipKernelNodeParams;
  ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2516:38: error: use of undeclared identifier 'cudaGraphKernelNodeGetParams'; did you mean 'hipGraphKernelNodeGetParams'?
                cudaError_t statRT = cudaGraphKernelNodeGetParams(nodes[i], &paramsRuntime[i]); // Get params using runtime
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                     hipGraphKernelNodeGetParams
C:/Program Files/AMD/ROCm/5.7/include\hip/hip_runtime_api.h:6607:12: note: 'hipGraphKernelNodeGetParams' declared here
hipError_t hipGraphKernelNodeGetParams(hipGraphNode_t node, hipKernelNodeParams* pNodeParams);
           ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2541:9: error: unknown type name 'cudaGraphExecUpdateResultInfo'; did you mean 'hipGraphExecUpdateResult'?
        cudaGraphExecUpdateResultInfo resultInfo;
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        hipGraphExecUpdateResult
C:/Program Files/AMD/ROCm/5.7/include\hip/hip_runtime_api.h:1199:3: note: 'hipGraphExecUpdateResult' declared here
} hipGraphExecUpdateResult;
  ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2545:20: error: use of undeclared identifier 'cudaGraphLaunch'; did you mean 'hipGraphLaunch'?
        CUDA_CHECK(cudaGraphLaunch(cudaGraph.instance, cuda_ctx->stream()));
                   ^~~~~~~~~~~~~~~
                   hipGraphLaunch
W:/git/test/agray3/llama.cpp/./ggml-cuda/common.cuh:186:40: note: expanded from macro 'CUDA_CHECK'
#define CUDA_CHECK(err) CUDA_CHECK_GEN(err, cudaSuccess, cudaGetErrorString)
                                       ^
W:/git/test/agray3/llama.cpp/./ggml-cuda/common.cuh:180:22: note: expanded from macro 'CUDA_CHECK_GEN'
        auto err_ = (err);                                                          \
                     ^
C:/Program Files/AMD/ROCm/5.7/include\hip/hip_runtime_api.h:6538:12: note: 'hipGraphLaunch' declared here
hipError_t hipGraphLaunch(hipGraphExec_t graphExec, hipStream_t stream);
           ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2816:62: warning: unused parameter 'buffer' [-Wunused-parameter]
GGML_CALL bool ggml_backend_cuda_register_host_buffer(void * buffer, size_t size) {
                                                             ^
W:/git/test/agray3/llama.cpp/ggml-cuda.cu:2816:77: warning: unused parameter 'size' [-Wunused-parameter]
GGML_CALL bool ggml_backend_cuda_register_host_buffer(void * buffer, size_t size) {
                                                                            ^
10 warnings and 9 errors generated when compiling for gfx1100.
ninja: build stopped: subcommand failed.
```

Engininja2 commented:
`nodes`, `paramsDriver`, and `paramsRuntime` are being used across multiple calls of the function but their data is only loaded in an earlier call. Should they be static?

agray3 committed:
FIx issues raised in comments

agray3 commented:
> `nodes`, `paramsDriver`, and `paramsRuntime` are being used across multiple calls of the function but their data is only loaded in an earlier call. Should they be static?

Good spot! I had this earlier when developing and somehow it got lost when refactoring/tidying but still worked. Now fixed.

agray3 commented:
@JohannesGaessler I think the llama-bench and perplexity issues should now be fixed with the latest commit - can you confirm from your end? Perplexity is slower with CUDA graphs ATM because is has batch size > 1 - as above I think I should only enable CUDA graphs for batch size 1 initially.

agray3 commented:
>It does not seems to work at P40 and I cannot get it compile on ROCM

>Tried to add ROCm HIP compatibility but it error 

The P40 issue may be due to the CUDA version in use, we need CUDA >= 12.0 for the functionality here. I have now added a macro such that this new code won't be compiled with earlier CUDA, and the original code will instead be used. Similarly it now won't be compiled with HIP/ROCm (which can be added in a follow up if there is adequate support and performance benefits on that platform).

agray3 committed:
Tidied to now only use CUDA runtime (not mixed with driver calls)

sorasoras commented:
> > It does not seems to work at P40 and I cannot get it compile on ROCM
> 
> > Tried to add ROCm HIP compatibility but it error
> 
> The P40 issue may be due to the CUDA version in use, we need CUDA >= 12.0 for the functionality here. I have now added a macro such that this new code won't be compiled with earlier CUDA, and the original code will instead be used. Similarly it now won't be compiled with HIP/ROCm (which can be added in a follow up if there is adequate support and performance benefits on that platform).

I am using 12.4cuda for my P40, but this might have something to do with hardware,right?

```
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 551.78                 Driver Version: 551.78         CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla P40                    WDDM  |   00000000:05:00.0 Off |                  Off |
| N/A   32C    P8             11W /  250W |       0MiB /  24576MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  Tesla P40                    WDDM  |   00000000:6F:00.0 Off |                  Off |
| N/A   26C    P8              9W /  250W |       0MiB /  24576MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```

sorasoras commented:
```
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 2 CUDA devices:
  Device 0: Tesla P40, compute capability 6.1, VMM: yes
  Device 1: Tesla P40, compute capability 6.1, VMM: yes
| model                          |       size |     params | backend    | ngl | sm         | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------- | ---------------: |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       | pp 512     |    435.30 ± 0.13 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       | tg 128     |     21.68 ± 0.02 |

build: 8960fe86 (2713)

ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 2 CUDA devices:
  Device 0: Tesla P40, compute capability 6.1, VMM: yes
  Device 1: Tesla P40, compute capability 6.1, VMM: yes
| model                          |       size |     params | backend    | ngl | sm         | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------- | ---------------: |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       | pp 512     |    435.63 ± 0.08 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       | tg 128     |     23.22 ± 0.05 |

build: 800f4fe (2701)

```
@agray3 ok,it seems work now.
there are decent improvement  on token generations.

agray3 commented:
@sorasoras thanks for testing. Can you let me know the exact command for which you are seeing a failure, so I can try and reproduce? I don't have access to P40 but I have done a test on P100 that works OK.

sorasoras commented:
> @sorasoras thanks for testing. Can you let me know the exact command for which you are seeing a failure, so I can try and reproduce? I don't have access to P40 but I have done a test on P100 that works OK.

I am running batch inference for translations.
It works for llama-bench but when I run it with server and inference via Openai

` server.exe -m W:\model\Qwen1.5_Q5KS.gguf -c 2048 -ngl 99 -cb --port 8081 -sm none -mg 0 -a Qwen_1.5_Q5KS
`

my guess is it has some issue with continuous batching

JohannesGaessler commented:
Some single GPU test results:

| GPU         | Model           | Test  | t/s master | t/s PR | Speedup |
|-------------|-----------------|-------|------------|--------|---------|
| 1x RTX 4090 | LLaMA 3 8b f16  | pp512 |       9776 |   9777 |    1.00 |
| 1x RTX 4090 | LLaMA 3 8b f16  | tg128 |      54.92 |  55.96 |    1.02 |
| 1x RTX 4090 | LLaMA 3 8b q4_0 | pp512 |       7387 |   7385 |    1.00 |
| 1x RTX 4090 | LLaMA 3 8b q4_0 | tg128 |     137.50 | 146.02 |    1.06 |
| 1x P40      | LLaMA 2 7b q4_0 | pp512 |        908 |    909 |    1.00 |
| 1x P40      | LLaMA 2 7b q4_0 | tg128 |      53.84 |  51.67 |    0.96 |
| 1x RX 6800  | LLaMA 2 7b q4_0 | pp512 |        712 |    713 |    1.00 |
| 1x RX 6800  | LLaMA 2 7b q4_0 | tg128 |      60.25 |  60.66 |    1.01 |

There is a performance regression for P40. I was not able to run multi GPU tests because `llama-bench` crashes. Error with 6x RTX 4090:

```
CUDA error: operation not supported
  current device: 2, in function ggml_backend_cuda_graph_compute at ggml-cuda.cu:2558
  cudaGraphInstantiate(&cudaGraph.instance, cudaGraph.graph, __null, __null, 0)
GGML_ASSERT: ggml-cuda.cu:60: !"CUDA error"
```

Error with 3x P40:

```
CUDA error: an illegal memory access was encountered
  current device: 2, in function ggml_backend_cuda_synchronize at ggml-cuda.cu:2403
  cudaStreamSynchronize(cuda_ctx->stream())
GGML_ASSERT: ggml-cuda.cu:60: !"CUDA error"
```

agray3 committed:
disable for multi-gpu and batch size > 1

sorasoras commented:
> With your changes, it now works with ROCm HIP (with patch below), but it is slower, making it likely not worth enabling it on that platform. I'm using a RX 6700 XT.
> 
> model	size	params	backend	ngl	test	t/s master	t/s PR	speedup
> llama 7B Q4_K - Medium	3.80 GiB	6.74 B	ROCm	99	pp 512	850.47 ± 0.18	850.52 ± 0.08	1.00
> llama 7B Q4_K - Medium	3.80 GiB	6.74 B	ROCm	99	tg 128	54.05 ± 0.02	52.86 ± 0.05	0.98
> The patch can be found here: https://gist.github.com/ardfork/a223a10d20961707e7b5f3ee0b76c7d5, didn't want to bloat your PR comments with a wall of text that will be useless for most people reading it.

It work decently on RDNA3 through.  I forget to record the value but it's about 2% faster

agray3 commented:
OK thanks. I've now disabled CUDA graphs for multi-GPU and batch size > 1 which should prevent these crashes and regressions (where I can investigate these cases later). I can also disable for Pascal, I'l have a look at that tomorrow (also assessing Volta etc).

agray3 committed:
Disable CUDA graphs for old GPU arch and with env var

jdecourval commented:
> > With your changes, it now works with ROCm HIP (with patch below), but it is slower, making it likely not worth enabling it on that platform. I'm using a RX 6700 XT.
> > model	size	params	backend	ngl	test	t/s master	t/s PR	speedup
> > llama 7B Q4_K - Medium	3.80 GiB	6.74 B	ROCm	99	pp 512	850.47 ± 0.18	850.52 ± 0.08	1.00
> > llama 7B Q4_K - Medium	3.80 GiB	6.74 B	ROCm	99	tg 128	54.05 ± 0.02	52.86 ± 0.05	0.98
> > The patch can be found here: https://gist.github.com/ardfork/a223a10d20961707e7b5f3ee0b76c7d5, didn't want to bloat your PR comments with a wall of text that will be useless for most people reading it.
> 
> It work decently on RDNA3 through. I forget to record the value but it's about 2% faster for token generation.

Here's on my 7900xtx

| model                          |       size |     params | backend    | ngl | test       |              t/s before | t/s after |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: | ------: |
| command-r 35B IQ3_XS - 3.3 bpw |  15.65 GiB |    37.08 B | ROCm       |  99 | pp 512     |    737.72 ± 1.57 | 738.31 ± 1.48 |
| command-r 35B IQ3_XS - 3.3 bpw |  15.65 GiB |    37.08 B | ROCm       |  99 | tg 128     |     29.00 ± 0.05 | 28.94 ± 0.02 |
| llama 8B Q5_K - Medium         |   4.78 GiB |     7.24 B | ROCm       |  99 | pp 512     |   3219.19 ± 8.14 | 3218.28 ± 8.52 |
| llama 8B Q5_K - Medium         |   4.78 GiB |     7.24 B | ROCm       |  99 | tg 128     |     87.13 ± 0.06 | 87.37 ± 0.06 |
| llama 8x7B IQ3_XXS - 3.0625 bpw |  33.27 GiB |    91.80 B | ROCm       |  99 | pp 512     |   1165.00 ± 4.31 | 1165.52 ± 4.11|
| llama 8x7B IQ3_XXS - 3.0625 bpw |  33.27 GiB |    91.80 B | ROCm       |  99 | tg 128     |     55.08 ± 0.04 | 55.39 ± 0.04 |
| llama 8B Q6_K                  |   5.53 GiB |     7.24 B | ROCm       |  99 | pp 512     |  3273.69 ± 10.66 | 3268.28 ± 10.24 |
| llama 8B Q6_K                  |   5.53 GiB |     7.24 B | ROCm       |  99 | tg 128     |     87.24 ± 0.07 | 87.51 ± 0.05 |

Over those 4 models, your PR yields a significant, but very small average TG speedup of 0.2%.

JohannesGaessler commented:
I just noticed that instead of using Github's built-in draft feature you added "DRAFT:" to the title. Please let me know when you think the PR is ready for review, I'll take a look then.

agray3 committed:
added missing CUDA_CHECKs

agray3 commented:
> I just noticed that instead of using Github's built-in draft feature you added "DRAFT:" to the title. Please let me know when you think the PR is ready for review, I'll take a look then.

Thanks @JohannesGaessler . I think it is ready now but it would be good to run the CI again. How do we do that? (The tests are automatically getting cancelled at the moment.) If that raises no issues then I'll remove DRAFT from the title (and use the built-in feature next time).

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+    if(cudaGraph.count==0){        
+        cudaDeviceProp prop;
+        int device;
+        CUDA_CHECK(cudaGetDevice(&device));
```
Use `ggml_cuda_get_device` and `ggml_cuda_set_device`. For getting the device this doesn't actually matter but when setting the device llama.cpp has a workaround for Windows to make the performance less bad compared to Linux.

Also I think it would be better to just get the device from `cuda_ctx`. Is that correct @slaren ?

agray3 replied:
Done (using cuda_ctx)

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+        cudaDeviceProp prop;
+        int device;
+        CUDA_CHECK(cudaGetDevice(&device));
+        CUDA_CHECK(cudaGetDeviceProperties(&prop, device));
+        if (prop.major < 8){
+            cudaGraph.disableDueToGpuArch=true;
+        }
```
Use `ggml_cuda_info().devices[device].cc` instead (same values as you would get with `__CUDA_ARCH__` in device code) and compare it to `CC_AMPERE` (800).

agray3 replied:
Done

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+        int k=0;
+        for (int i = 0; i < cgraph->n_nodes; i++) {
+            ggml_tensor * node = cgraph->nodes[i];
+            // Identify if the graph needs updated for this token due to the number of elements changing
```
```suggestion
            // Identify if the graph needs to be updated for this token due to the number of elements changing
```

agray3 replied:
Done

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+    bool cudaGraphUpdateRequired = false;
+#endif
+    
+    // Only perfom the graph exection if CUDA graphs are not enebled, or we are capturing the graph.
+    // With use of CUDA graphs, the execution will be performed by the graph launch.
```
```suggestion
    // Only perform the graph execution if CUDA graphs are not enabled, or we are capturing the graph.
    // With the use of CUDA graphs, the execution will be performed by the graph launch.
```

agray3 replied:
Done

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+#else
+    bool useCudaGraph = false;
+    bool cudaGraphUpdateRequired = false;
+#endif
```
```suggestion
#endif // USE_CUDA_GRAPH
```

agray3 replied:
Done (and similar at the end of the function)

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
     }
+    }
+
+    #ifdef USE_CUDA_GRAPH
```
```suggestion
#ifdef USE_CUDA_GRAPH
```

agray3 replied:
Done

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
+            }
+            CUDA_CHECK(cudaGraphGetNodes(cudaGraph.graph, cudaGraph.nodes, &cudaGraph.numNodes));
+
+            // Loop over nodes, and extract kernel parameters fro each node
```
```suggestion
            // Loop over nodes, and extract kernel parameters from each node
```

agray3 replied:
Done

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
 
+        if(cudaGraphUpdateRequired) {
+            // Extract nodes from graph
+            if(cudaGraph.numNodes == 0) {
+                CUDA_CHECK(cudaGraphGetNodes(cudaGraph.graph, nullptr, &cudaGraph.numNodes));
+            }
+            CUDA_CHECK(cudaGraphGetNodes(cudaGraph.graph, cudaGraph.nodes, &cudaGraph.numNodes));
```
Why is `cudaGraphGetNodes` being called twice for `cudaGraph.numNodes == 0`?

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
+                CUDA_CHECK(cudaGraphNodeGetType(cudaGraph.nodes[i], &nodeType));
+                if (nodeType == cudaGraphNodeTypeKernel) {
+                    auto statRT = cudaGraphKernelNodeGetParams(cudaGraph.nodes[i], &cudaGraph.params[i]); // Get params using runtime
+                    if(statRT == cudaErrorInvalidDeviceFunction) {
+                        // Fails due to incorrect handling by CUDA runtime of CUDA BLAS node.
+                        // We don't need to update blas nodes, so clear error and move on.
+                        cudaGetLastError();
+                    }
```
Isn't this error catching too broad? Can this result in bad behavior that is very difficult to debug?

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
+            }
+        }
+
+        // Update copy kernel param (required every token)
```
I'm not sure what this comment is supposed to say. "requires every token" or "required by every token"?

agray3 replied:
I've now made this clearer:

        // One of the arguments to the copy kernel is updated for each token, hence we need to
        // replace that argument with the updated value in the CUDA graph

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
+        // Update graph executable
+        cudaGraphExecUpdateResultInfo resultInfo;
+        auto stat = cudaGraphExecUpdate(cudaGraph.instance, cudaGraph.graph, &resultInfo);
+        if(stat == cudaErrorGraphExecUpdateFailure)
+        {
```
`{` should be on the same line as `if (...)`

agray3 replied:
Done

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -2432,7 +2521,71 @@ GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t
+        if(stat == cudaErrorGraphExecUpdateFailure)
+        {
+            // The pre-existing graph exec cannot be updated due to violated constraints
+            // so instead clar error and re-instantiate
```
```suggestion
            // so instead clear error and re-instantiate
```

JohannesGaessler reviewed (commented):
Many of my comments are just about code style, mostly snake_case vs. CamelCase and `int* var` vs. `int * var`, see https://github.com/ggerganov/llama.cpp?tab=readme-ov-file#coding-guidelines (I didn't flag every single instance of those).

Also I think a review from @slaren would make sense since compared to me he worked on the CUDA host code more recently.

My RTX 4090 machine is currently busy with importance matrix calculation, I will do more testing on Saturday at the latest.

agray3 committed:
Addressed comments

agray3 commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+    int softmax_ne0 = 0;
+    cudaGraphNode_t nodes[MAX_NODES_IN_CUDA_GRAPH];
+    cudaKernelNodeParams params[MAX_NODES_IN_CUDA_GRAPH];
+    bool disableDueToGpuArch=false;
```
Done

agray3 commented on ggml-cuda.cu:
```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+    // Objects required for CUDA Graph
+    static ggml_cudaGraph cudaGraph;
+    bool useCudaGraph = (cudaGraph.count>=7); //avoid CUDA graphs on first few steps due to incompatible initialisations.
+    char** updatedKernelArg[MAX_NODES_IN_CUDA_GRAPH];
```
Done

agray3 committed:
further addressed comments

slaren commented:
As it is, this is going to break every user of ggml-cuda other than llama.cpp, and even within llama.cpp this approach will fail to detect changes to the graph in some cases, which will lead to hard to debug issues. The implementation makes too many assumptions about the graphs that it is going to receive and it will break very easily. For instance, this breaks MoE models because mul_mat_id does some operations that cannot be recorded into a graph. `test-backend-ops` breaks on the second op. In my opinion, the implementation needs to be a lot more robust before it can merged.

FWIW, this is precisely the use case that the `graph_plan_*` functions of the ggml-backend interface are meant to cover. It would need some adjustments to be able to update graph plans, and it would need cooperation from the application side, but with that we would have a path to implement this in a more robust and reusable way.

Still, I appreciate that the performance improvement is very significant, and it would be very desirable to have this capability on llama.cpp as soon as possible, so it may be worth to make an effort to solve the issues with this approach and merge it as a llama.cpp-only temporary solution. In this case, we would need to make more checks, for example we may want to keep a copy of the `ggml_cgraph` used to record the CUDA graph, and check node by node that all the operations are the same and that the tensors have the same addresses. The implementation should also be gated behind a compile time flag that is only enabled in llama.cpp.

ggerganov commented:
> In my opinion, the implementation needs to be a lot more robust before it can merged.

> it would be very desirable to have this capability on llama.cpp as soon as possible, so it may be worth to make an effort to solve the issues with this approach and merge it as a llama.cpp-only temporary solution.

I agree and to add to this - speed improvements are obviously welcome, but we should be mindful about compatibility issues and long-term support. In that sense, there is no rush to merge - we can take some time if needed to consider the best way to integrate this functionality. A temporary solution, gated with a compile flag is also an option as proposed.

@slaren Feel free to propose any improvements and a suitable path for integration. Your input is very appreciated.

agray3 committed:
limit to GGML_ALLOW_CUDA_GRAPHS defined in llama.cpp cmake

agray3 commented:
Thanks @slaren and @ggerganov . I'm not an expert on all the different usage possibilities so I appreciate the guidance, and more than happy to further improve robustness. I've now added the compilation limitation as suggested (new GGML_ALLOW_CUDA_GRAPHS macro defined in llama.cpp CMakelists.txt). 

> for example we may want to keep a copy of the ggml_cgraph used to record the CUDA graph, and check node by node that all the operations are the same and that the tensors have the same addresses.

Note that there is not a 1-1 correspondence between the nodes in the ggml_cgraph and CUDA graph. The CUDA graph has 1 node per CUDA kernel; the ggml_cgraph has some nodes without any kernels and some nodes that launch multiple kernels. However I think it should be possible to detect whenever the graph capture fails and automatically fall back. There will be a limited number of CUDA operations in llama.cpp that can potentially cause such failures, so we can add suitable error handling to each of these. I'll start looking at this and report back.

slaren commented:
The main issue is that it cannot be assumed that all the graphs that `ggml_backend_cuda_graph_compute` receives are the same. Even within `llama.cpp`, the graphs will vary depending on the parameters of the evaluation. In some cases there are completely different graphs being evaluated, eg. for KV defragmentation. Even if the graph seemingly is the same, the tensors may have been reallocated and their addresses changed. So the logic to determine if the CUDA graph can be reused needs to be a lot more robust. One option would be to keep a copy of the `ggml_cgraph` used to create the CUDA graph, and compare the graph being evaluated to that. That's not entirely trivial because cloning `ggml_cgraph` is not very straightforward, so instead it may be easier to keep a simplified representation of the graph that keeps a list of nodes, their address, the op, and the addresses of the sources.

agray3 commented:
>The main issue is that it cannot be assumed that all the graphs that ggml_backend_cuda_graph_compute receives are the same. Even within llama.cpp, the graphs will vary depending on the parameters of the evaluation. In some cases there are completely different graphs being evaluated, eg. for KV defragmentation. Even if the graph seemingly is the same, the tensors may have been reallocated and their addresses changed. So the logic to determine if the CUDA graph can be reused needs to be a lot more robust. One option would be to keep a copy of the ggml_cgraph used to create the CUDA graph, and compare the graph being evaluated to that. That's not entirely trivial because cloning ggml_cgraph is not very straightforward, so instead it may be easier to keep a simplified representation of the graph that keeps a list of nodes, their address, the op, and the addresses of the sources.

Got it. Yes, that makes a lot of sense - I'll go ahead and add this checking as you suggest.

JohannesGaessler commented:
How about this: in terms of performance I think it would make sense to double buffer ggml graph creation. So essentially start creating the next ggml graph while the previous one is still being evaluated. If you then have these two graphs you could implement a function that checks whether the two graphs are equivalent and use that information to decide whether or not you need to rebuild the CUDA graph.

agray3 commented:
> How about this: in terms of performance I think it would make sense to double buffer ggml graph creation. So essentially start creating the next ggml graph while the previous one is still being evaluated. If you then have these two graphs you could implement a function that checks whether the two graphs are equivalent and use that information to decide whether or not you need to rebuild the CUDA graph.

This solution would have the added benefit of overlapping the ggml graph creation with execution of the previous token so would further improve performance (as discussed on #6763). I can see from the profile that there is a Device to Host memcpy after each graph execution (128kB), which I assume is the generated token. Are we sure that there is not a dependency on that data on the creation of the next ggml graph?

agray3 committed:
Merge branch 'ggerganov:master' into ag_cuda_graphs

slaren commented:
> I can see from the profile that there is a Device to Host memcpy after each graph execution (128kB), which I assume is the generated token. Are we sure that there is not a dependency on that data on the creation of the next ggml graph?

The process is usually evaluate token -> synchronize -> copy logits to CPU -> sample a token on the CPU side -> evaluate next token. The topology of the graph does not depend on the sampled token, the token(s) is in an input tensor copied to VRAM before evaluation begins. So, in principle, it is possible to prepare the next graph before the current evaluation is completed, but the API is not really setup to support that use case very well. IMO it is not worth to bother with this at this point, the time spent creating the graph is not significant for the model sizes that we typically use. Doing so could improve performance with very small models, but that's not really a priority at the moment. Simply adding a comprehensive check to ensure that the graph recorded in the CUDA graph is the same as the graph being evaluated should be more than good enough for now.

agray3 committed:
Added more comprehensive graph node checking

agray3 committed:
Fall back if graph capture fails and address other comments

agray3 committed:
With mechanism to fall back if graph capture fails

agray3 commented:
I've now also leveraged the existing CUDA_CHECK/ggml_cuda_error() to implement a mechanism to allow automatic fallback if CUDA graph capture fails.

agray3 committed:
Revert "With mechanism to fall back if graph capture fails"

This reverts commit eb9f15fb6fcb81384f732c4601a5b25c016a5143.

agray3 commented:
> I've now also leveraged the existing CUDA_CHECK/ggml_cuda_error() to implement a mechanism to allow automatic fallback if CUDA graph capture fails.

Something not quite right with this, reverting it for now

slaren commented:
About the reverted CUDA graph capture failure check, it would be better to modify the `CUDA_CHECK` macro and check the error code directly instead of the string. Another option may be to check the graph nodes for the op `GGML_OP_MUL_MAT_ID`, I think that's the only one that cannot be captured.

agray3 commented:
- I've now made the provided change, thanks, and yes it was necessary to exempt `GGML_OP_CPY` and also `GGML_OP_VIEW` (which doesn't contribute any activities to the graph anyway).
- I've explicitly disabled CUDA graphs in the presence of any `GGML_OP_MUL_MAT_ID` node
- I've moved the `cuda_graph` object to `ggml_backend_cuda_context` as suggested.
- I've re-enabled (and fixed) the fallback mechanism. It is still using strings in `ggml_cuda_error()` because I need it to set the `disable_cuda_graphs_due_to_failed_capture` flag in `ggml-cuda.cu`  - I'm not sure how to do that cleanly using the macro. But this mechanism may be superfluous now given the GGML_OP_MUL_MAT_ID handling above - let me know if you just prefer it to be removed.

agray3 committed:
Merge branch 'ggerganov:master' into ag_cuda_graphs

slaren commented:
Thanks @agray3, I think this should work. I will take a look at this and cleanup the code style and other minor issues when I have the chance.

slaren committed:
Merge remote-tracking branch 'origin/master' into ag_cuda_graphs

slaren committed:
- renamed GGML_ALLOW_CUDA_GRAPHS to GGML_CUDA_USE_GRAPHS

- rename env variable to disable CUDA graphs to GGML_CUDA_DISABLE_GRAPHS

- updated Makefile build to enable CUDA graphs

- removed graph capture failure checking in ggml_cuda_error
  using a global variable to track this is not thread safe, but I am also not safistied with checking an error by string
  if this is necessary to workaround some issues with graph capture with eg. cuBLAS, we can pass the ggml_backend_cuda_context to the error checking macro and store the result in the context

- fixed several resource leaks

- fixed issue with zero node graphs

- changed fixed size arrays to vectors

- removed the count of number of evaluations before start capturing, and instead changed the capture mode to relaxed

- removed the check for multiple devices so that it is still possible to use a single device, instead checks for split buffers to disable cuda graphs with -sm row

- changed the op for checking batch size to GGML_OP_ADD, should be more reliable than GGML_OP_SOFT_MAX

- code style fixes

- things to look into
  - VRAM usage of the cudaGraphExec_t, if it is significant we may need to make it optional
  - possibility of using cudaStreamBeginCaptureToGraph to keep track of which ggml graph nodes correspond to which cuda graph nodes

slaren commented:
Fixed all the issues I could find, I think this is ready to be merged, but more testing would be welcome.

I get this performance on my system (with flash attention enabled):

| GPU         | Model          |   Model Size [GiB] |  Test   |   t/s master |   t/s ag/ag_cuda_graphs |   Speedup |
|:------------|:---------------|-------------------:| :-------|-------------:|------------------------:|----------:|
| RTX 3090 Ti | llama 7B F16   |              12.55 |  tg128  |        57.05 |                   59.80 |      1.05 |
| RTX 3090 Ti | llama 7B Q4_0  |               3.56 |  tg128  |       138.58 |                  163.95 |      1.18 |
| RTX 3090 Ti | llama 13B Q4_0 |               6.86 |  tg128  |        85.43 |                   95.73 |      1.12 |
| RTX 3090 Ti | llama 30B Q4_0 |              17.09 |  tg128  |        39.32 |                   42.72 |      1.09 |

@agray3 I was wondering if it would be possible to use `cudaStreamBeginCaptureToGraph` to keep track of which nodes in the ggml graph correspond to which nodes in the cuda graph. In other words, is it possible to query the nodes of the cuda graph as they are captured, or only after `cudaStreamEndCapture`? If so, I think this could allow us to implement a more reliable update mechanism in the future.

slaren committed:
fix build without cuda graphs

agray3 commented:
>I was wondering if it would be possible to use cudaStreamBeginCaptureToGraph to keep track of which nodes in the ggml graph correspond to which nodes in the cuda graph. In other words, is it possible to query the nodes of the cuda graph as they are captured, or only after cudaStreamEndCapture? If so, I think this could allow us to implement a more reliable update mechanism in the future.

Unfortunately the graph is not exposed in the app until cudaStreamEndCapture(). You could possibly do something like this manually by introducing a wrapper to all kernel launches, plus some mechanism in that wrapper to keep track of things.

ggerganov commented on ggml-cuda.cu:
```diff
@@ -2410,32 +2410,306 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+            if (!has_matching_properties) {
+                cuda_graph_update_required = true;
+            }
+            set_ggml_graph_node_properties(cgraph->nodes[i], &cuda_ctx->cuda_graph->ggml_graph_properties[i]);
```
Do we need to re-set the properties if `has_matching_properties == true`?

slaren replied:
I think so, because once `cuda_graph_update_required` is set, `has_matching_properties` is always `true`.

agray3 replied:
Yes, `set_ggml_graph_node_properties()` is required even when `has_matching_properties == true` because these properties need to be compared against in the following token.

slaren committed:
remove outdated comment

slaren committed:
replace minimum cc value with a constant

slaren commented:
> Unfortunately the graph is not exposed in the app until cudaStreamEndCapture(). You could possibly do something like this manually by introducing a wrapper to all kernel launches, plus some mechanism in that wrapper to keep track of things.

If I understand `cudaStreamBeginCaptureToGraph` correctly, I imagine it would be possible to call `cudaStreamEndCapture` after every op, take note of the current number of nodes, and then restart the capture using `cudaStreamBeginCaptureToGraph`. Probably wouldn't be very efficient, though.

JohannesGaessler reviewed (commented):
Based on static code analysis I only have minor comments like `i++` vs. `++i` and the things I otherwise mentioned. Definitely nothing that I would consider worth blocking a merge for; Let me also check the performance and correctness.

agray3 commented:
>If I understand cudaStreamBeginCaptureToGraph correctly, I imagine it would be possible to call cudaStreamEndCapture after every op, take note of the current number of nodes, and then restart the capture using cudaStreamBeginCaptureToGraph. Probably wouldn't be very efficient, though. It's not very clear from the documentation if cudaStreamBeginCaptureToGraph clears the graph, or requires an empty graph, or appends to it.

Yes, you are right, that should work - I've not tried it but it should append. But as you say I think it would have quite substantial overhead because there are a lot of kernels in each graph.

slaren commented on ggml-cuda.cu:
```diff
@@ -2410,32 +2410,306 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+    bool graph_evaluated_or_captured = false;
+
+    while (!graph_evaluated_or_captured) {
+        // Temporarily avoid indenting here (and below the following if) to make code review easier
```
Yep, already removed.

slaren commented on ggml-cuda.cu:
```diff
@@ -2410,32 +2410,306 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+                assert(node->buffer->buft == ggml_backend_cuda_buffer_type(cuda_ctx->device));
+                for (int j = 0; j < GGML_MAX_SRC; j++) {
+                    if (node->src[j] != nullptr) {
+                        assert(node->src[j]->buffer->buft == ggml_backend_cuda_buffer_type(cuda_ctx->device) || ggml_backend_buffer_is_cuda_split(node->src[j]->buffer));
```
These are correctness checks for ggml-backend that are only meant to be enabled in debug builds (that code was already before). In normal circumstances this should never happen, so there is no need to check it on release builds.

slaren commented on ggml-cuda.cu:
```diff
@@ -2410,32 +2410,306 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+            } else {
+                graph_evaluated_or_captured = true; // CUDA graph has been captured
+            }
+#endif
```
This code used to check for graph capture failures, but I removed that since the implementation had issues. I think that there shouldn't be any capture failures if the ggml graph is checked correctly for incompatible ops early on. Also it uses relaxed capture mode now, which allows ops such as allocations to succeed even if they cannot be captures into the graph.

JohannesGaessler reviewed (approved):
| GPU      | Model         |   Model Size [GiB] | Test   |   t/s master |   t/s f42312e0 |   Speedup |
|:---------|:--------------|-------------------:|:-------|-------------:|---------------:|----------:|
| RTX 4090 | llama 8B F16  |              14.96 | tg128  |        56.68 |          57.85 |      1.02 |
| RTX 4090 | llama 8B Q4_0 |               4.33 | tg128  |       149.28 |         159.67 |      1.07 |

According to my testing the results are bit-for-bit identical, the evaluation is simply faster. It's unfortunate that it right now only works for batch size 1 since that means it won't be compatible with >1 server slots or speculative decoding such as with https://github.com/ggerganov/llama.cpp/pull/6828 (I'm not sure whether the constant switching between batch sizes causes issues but I consider that my own problem to fix).

ggerganov commented:
I don't have Ampere hardware handy, but the PR builds and works on RTX 2060 as usual

fat-tire commented:
> I don't have Ampere hardware handy, but the PR builds and works on RTX 2060 as usual

I'm guessing 30x0 is good too (also Ampere)...

JohannesGaessler commented:
Did anyone check server performance before and after this PR? I am seeing no difference in terms of user request throughput on 1x RTX 4090 even with a single server slot: https://github.com/ggerganov/llama.cpp/pull/6828#issuecomment-2102746433

slaren commented:
If it is being disabled for some reason, you can find out with a debug build.

JohannesGaessler commented:
According to the log with debugging enabled this seems to be the reason:

`
ggml_backend_cuda_graph_compute: disabling CUDA graphs due to batch size > 1 [ffn_inp-0] [4096 256 1 1]
ggml_backend_cuda_graph_compute: disabling CUDA graphs due to batch size > 1 [ffn_inp-0] [4096 256 1 1]
ggml_backend_cuda_graph_compute: disabling CUDA graphs due to batch size > 1 [ffn_inp-0] [4096 256 1 1]
ggml_backend_cuda_graph_compute: disabling CUDA graphs due to batch size > 1 [ffn_inp-0] [4096 256 1 1]
ggml_backend_cuda_graph_compute: disabling CUDA graphs due to too many consecutive updates
`

I think the problem is that the graph is initially `NULL` so this sets `cuda_graph_update_required = true` and because the microbatches are > 1 this also doesn't set the graph to not `NULL`. As a consequence a sufficiently long prompt triggers `disable_due_to_too_many_updates = true` and the graphs are permanently disabled.

slaren commented:
The logic to determine when to disable graphs is not great, but necessary to avoid degrading performance when cuda graphs cannot be used. For example with pipeline parallelism tensor addresses change with every eval so it must be disabled. But maybe prompt evals shouldn't increase the counter.

agray3 commented:
@JohannesGaessler can you let me know how to reproduce this? I can then try to relax these checks appropriately. Thanks.

agray3 commented:
Thanks. Yeah I confirm it works if I comment out the line
``` cuda_ctx->cuda_graph->number_consecutive_updates++;```
I just need to work out the best way to stop this counter incrementing due to long prompts - will take a look next week.

agray3 commented:
> Thanks. Yeah I confirm it works if I comment out the line ` cuda_ctx->cuda_graph->number_consecutive_updates++;` I just need to work out the best way to stop this counter incrementing due to long prompts - will take a look next week.

From playing around with the code, it seems that, in `main.cpp`, we know the prompt is still evaluating when `n_past==0` so could use that information to disable CUDA graphs while the prompt is still evaluating. However the question is how to propagate that info down to the ggml-cuda module, since there are a several layers in between. Firstly, @JohannesGaessler  @slaren do you know if there is another way to assess if the prompt is evaluating at a deeper level? If not, then I think it will require a series of API calls through the layers to set/unset a flag in ggml-cuda.

slaren commented:
I think you can use the already existing code that checks for batch size > 1 to do this. I changed your implementation to check for add operation instead of a softmax, but the idea is the same. Did you find it unreliable?

llama.cpp knows when it is a prompt (or batch) evaluation when the call to `llama_decode` has `batch.n_tokens > 1`. The way I would like to do this in the long run is by implementing the `graph_plan` functions of the ggml-backend interface using cuda graphs, and have llama.cpp use a graph plan for single token batches when possible (we would also need to add a `graph_plan_update` function, there isn't one currently).

agray3 commented:
Thanks, I found the issue - the counter is being unnecessarily incremented even for tokens where graphs are disabled. See the simple fix at https://github.com/ggerganov/llama.cpp/pull/7302
